File: LINKLIST/JAVA/singlyCLL.java
class node
{
    public int data ;
    public node next;
}

class singlyCLLX
{
    private node first;
    private node last ;
    private int icount;

    public
    singlyCLLX()
    {
        first = null;
        last = null;
        icount = 0;
    }

    void insertfirst(int no)
    {
        node newn = null;
        newn = new node ();

        newn.data = no;
        newn.next = null;

        if(first == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }
        last.next = first;
        icount++;
    }

    void insertlast(int no)
    {
        node newn = null;
        newn = new node ();

        newn.data = no;
        newn.next = null;

        if(first == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            last = newn;
        }
        last.next = first;
        icount++;
    }

    void insertatpos(int no,int pos)
    {
        node newn = null;
        node temp = null;
        int icnt = 0;

        icount = count();

        if(pos < 1 || pos > icount + 1)
        {
            System.out.println("invalid position");
            return;
        }

        if(pos == 1)
        {
            insertfirst(no);
        }
        else if(pos == icount + 1)
        {
            insertlast(no);
        }
        else
        {
            temp = first;
            newn = new node();

            newn.data = no;
            newn.next = null;

            for(icnt = 1;icnt < pos-1 ;icnt++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next= newn;

            icount++;

        }

    }

    void deletefirst()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if (first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;
        }
        last.next= first;
        icount--;
    }

    void deletelast()
    {
        if(first == null && last == null)
        {
            return;
        }
        else if (first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            node temp = null;
            temp = first;

            while(temp.next.next != first)
            {
                temp = temp.next;
            }
            temp.next = first;
            last = temp;
        }
        last.next = first;
        icount--;
    }

    void deleteatpos(int pos)
    {
        node temp = null;
        int icnt = 0;
        icount = count();

        if(pos < 1 || pos > icount )
        {
            System.out.println("invalid position");
            return;
        }

        if(pos == 1)
        {
            deletefirst();
        }
        else if(pos == icount )
        {
            deletelast();
        }
        else
        {
            temp = first;

            for(icnt = 1;icnt < pos - 1;icnt++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;

            icount--;
        }
    }

    void display()
    {
        node temp = null;

        if(first == null && last == null)
        {
            return;
        }
        temp = first;

        System.out.print("->");
        do
        {
            System.out.print(" | "+temp.data+" | -> ");
            temp = temp.next;
        }while(temp != last.next);
        System.out.println();
        
    }

    int count ()
    {
        return icount;
    }

}

class singlyCLL
{
    public static void main(String a[])
    {
        singlyCLLX sobj = new singlyCLLX();
        int iret = 0;

        sobj.insertfirst(51);
        sobj.insertfirst(21);
        sobj.insertfirst(11);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.insertlast(101);
        sobj.insertlast(111);
        sobj.insertlast(121);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.insertatpos(105,5);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);


        sobj.deletefirst();

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.deletelast();

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.deleteatpos(5);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

    }
    
}

File: singlyCLL.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node * next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void insertfirst(PPNODE first,PPNODE last,int no)
{
    PNODE newn = NULL;
    newn = (PNODE)malloc(sizeof(NODE));

    newn -> data = no;
    newn -> next = NULL;

    if(*first == NULL && *last == NULL)
    {
       *first = newn;
       *last = newn;
       
       (*last)->next= *first;
    }
    else
    {
        newn -> next = *first;
        *first = newn;

        (*last)-> next = *first;
    }

}

void insertlast(PPNODE first,PPNODE last,int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn -> data = no;
    newn-> next = NULL;

    if(*first == NULL && *last == NULL)
    {
        *first = newn;
        *last = newn;

        (*last)-> next = *first;

    }
    else
    {
        (*last)-> next = newn;
        *last = newn;

        (*last) -> next  = *first;

    }
}

void deletefirst(PPNODE first,PPNODE last )
{
    PNODE temp = NULL;

    if(*first == NULL && *last == NULL)
    {
        return;
    }
    else if (*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;

    }
    else
    {
        temp = *first;
        *first = temp -> next;

        free(temp);
        (*last)-> next = *first;
    }
}

void deletelast(PPNODE first,PPNODE last)
{
    PNODE temp = NULL;

    if(*first == NULL && *last == NULL)
    {
        return;
    }
    else if(*first == *last)
    {
        free(*last);
        *first = NULL;
        *last = NULL;
    }
    else
    {
        temp = *first;
        while(temp-> next ->next != *first)
        {
            temp = temp->next;
        }
        *last = temp;
        free(temp->next);

        (*last)->next = *first;


    }
}

void display(PNODE first,PNODE last)
{
    if(first == NULL && last == NULL)
    {
        return;
    }
    do
    {
        printf("| %d | -> ",first->data);
        first= first->next;
    }while(first != last-> next);
    printf("\n");
}

int count(PNODE first,PNODE last)
{
    int icount = 0;
    if(first == NULL && last == NULL)
    {
        return icount; 
    }

    do
    {
        icount++;
        first = first ->next;
    }while(first != last->next);
    return icount;
}

void insertatpos(PPNODE first,PPNODE last,int no,int pos)
{
    PNODE temp = NULL;
    PNODE newn = NULL;
    int icount = 0;
    int icnt = 0;

    icount = count(*first,*last);

    if(pos < 1 || pos > icount+1)
    {
        printf("invalid position entered \n");
        return;
    }

    if(pos == 1)
    {
        insertfirst(first,last,no);
    }
    else if(pos == icount+1)
    {
        insertlast(first,last,no);
    }
    else
    {
        temp = *first;

        newn = (PNODE)malloc(sizeof(NODE));

        newn ->data = no;
        newn->next = NULL;

        for(icnt = 1;icnt < pos-1;icnt++)
        {
            temp = temp->next;
        }
        newn->next = temp ->next;
        temp ->next = newn;
        
    }
}

void deleteatpos(PPNODE first,PPNODE last,int pos)
{
    PNODE temp = NULL;
    PNODE target = NULL;
    int icount = 0;
    int icnt = 0;

    icount = count(*first,*last);

    if (pos < 1 || pos > icount )
    {
        printf("invalid syntax\n");
        return ;
    }

    if(pos == 1)
    {
        deletefirst(first,last);
    }
    else if(pos == icount)
    {
        deletelast(first,last);
    }
    else
    {
        temp = *first;

        for(icnt = 1;icnt < pos-1;icnt++)
        {
            temp = temp ->next;
        }

        target = temp->next;
        temp ->next = target ->next;
        free(target);

    }
}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;
    int iret = 0;

    insertfirst(&head,&tail,51);
    insertfirst(&head,&tail,21);
    insertfirst(&head,&tail,11);

    display(head,tail);

    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    insertlast(&head,&tail,101);
    insertlast(&head,&tail,111);
    insertlast(&head,&tail,121);

    display(head,tail);

    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    deletefirst(&head,&tail);

    display(head,tail);

    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    deletelast(&head,&tail);
    display(head,tail);

    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    insertatpos(&head,&tail,105,5);

    display(head,tail);

    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);

    deleteatpos(&head,&tail,5);

    display(head,tail);

    iret = count(head,tail);
    printf("no of element in linklist is :%d\n",iret);


    return 0;
}